src/deco/deco.hpp

```diff
@@ -213,66 +212,120 @@ class ViewDecoration : public wf::compositor_surface_t,
         &_subsurfs.bottom_left, &_subsurfs.bottom_right,
     };
 
-    int width = 0;  ///< Width of the decoration.
-    int height = 0; ///< Height of the decoration.
+    wf::dimensions_t size; ///< Size of the decoration.
 
     wf::region_t cached_region; ///< Cached minimal region containing this deco.
 
-    /// Calculate the minimal region that contains this decoration surface.
-    wf::region_t calculate_region();
+  public:
+    DecorationSurface(ViewNodeRef node, nonstd::observer_ptr<Options> options)
+        : node(node), options(options) {}
+
+    // Set the size of the surface.
+    void set_size(wf::dimensions_t view_size);
+
+    // Set the surface color as active or inactive.
+    void set_active(bool active);
+
+    // Recalculate the region and cache it.
+    void recalculate_region();
+
+    // Unmap the surface.
+    void unmap();
+
+    // == Impl wf::surface_interface_t ==
+    bool is_mapped() const override;
+    wf::point_t get_offset() override;
+    wf::dimensions_t get_size() const override;
+    bool accepts_input(int32_t sx, int32_t sy) override;
+    void simple_render(const wf::framebuffer_t &fb, int x, int y,
+                       const wf::region_t &damage) override;
+};
+
+class ViewDecoration : public wf::decorator_frame_t_t {
+  private:
+    const ViewNodeRef node; ///< The node we're decorating.
+
+    /// Surface representing the decoration.
+    nonstd::observer_ptr<DecorationSurface> surface_ref;
+
+    /// Surface swap, used when hiding the surface from the node.
+    std::unique_ptr<wf::surface_interface_t> surface;
+
+    /// The loaded options from the cfg.
+    nonstd::observer_ptr<Options> options;
 
     wf::signal_connection_t on_prefered_split_type_changed =
         [&](wf::signal_data_t *) { damage(); };
 
     wf::signal_connection_t on_config_changed = [&](wf::signal_data_t *) {
         // Refresh geometry in case border_width changes.
         node->refresh_geometry();
-
-        // Refreshing the geometry may not actually change the geometry. (e.g.
-        // If only border_radius changes) So we still need to update the
-        // cached_region here.
-        cached_region = calculate_region();
+        surface_ref->recalculate_region();
         node->view->damage();
     };
 
     wf::signal_connection_t on_detached = [&](wf::signal_data_t *) {
-        mapped = false;
-        wf::emit_map_state_change(this);
-        node->view->set_decoration(nullptr); // ViewDecoration dies here.
+        surface_ref->unmap();
+        detach_surface();
```

Comment:
To prevent an assertion error:
```suggestion
        if (!is_hidden())
            detach_surface();
```